### PR TITLE
Including metrics provider services

### PR DESCRIPTION
**Pull request by mdelosdolores-vf**

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.stream.StreamSupport;
+import javax.inject.Inject;
+
+public class ProgramStatusMetricsProvider implements MetricsProvider {
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java]: Name it SparkProgramStatusMetricProvider and also check for the program type being Spark.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+  @Override
+  public ExecutionMetrics retrieveMetrics(ProgramRunId runId) {
+    ExecutionMetrics metrics = null;
+    int retriesCount = 0;
+    String runIdStr = runId.getRun();
+    String sparkHistoricBaseURL = cConf.get(SPARK_BASE_URL_CONFIGURATION);
+    String applicationsURL = String.format("%s%s", sparkHistoricBaseURL,
+                                           SPARK_APPLICATIONS_ENDPOINT);
+    while (Objects.isNull(metrics) && retriesCount < MAX_RETIRES) {
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java]: Use https://github.com/cdapio/cdap/blob/develop/cdap-common/src/main/java/io/cdap/cdap/common/service/Retries.java for retry

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  @VisibleForTesting
+  protected String extractAttemptId(String responseBody, String runId) {
+    final String[] attemptId = new String[1];
+    JsonArray jsonAppResponse = jsonParser.parse(responseBody)
+      .getAsJsonArray();
+
+    StreamSupport.stream(jsonAppResponse.spliterator(), false)
+      .filter(app -> runId.equals(app.getAsJsonObject().get("id").getAsString()))
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java]: Is it possible to filter with application Id in the query itself?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      .findFirst().ifPresent(app -> {
+        JsonArray attArray = app.getAsJsonObject().get("attempts").getAsJsonArray();
+        StreamSupport.stream(attArray.spliterator(), false)
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java]: Can the response be mapped to a Java object using Gson?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java (hunk 3 above)
Comment (mdelosdolores-vf) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusMetricsProvider.java]: No, but we can query for programs ended in a 5 minutes range, this way the response of the request is lighter.

**Pull request closed without merging**